File: frontend/src/app/api/folders.ts
import { apiClient } from './client'
import { Folder, FolderItem } from '../types'

export const getFolders = async (): Promise<Folder[]> => {
  const res = await apiClient.get<Folder[]>('/api/folders')
  return res.data
}

export const createFolder = async (name: string): Promise<Folder> => {
  const res = await apiClient.post<Folder>('/api/folders', { name })
  return res.data
}

export const renameFolder = async (folderId: string, name: string): Promise<Folder> => {
  const res = await apiClient.patch<Folder>(`/api/folders/${folderId}`, { name })
  return res.data
}

export const deleteFolder = async (folderId: string): Promise<void> => {
  await apiClient.delete(`/api/folders/${folderId}`)
}

export const addItemToFolder = async (
  folderId: string,
  item: {
    content_id: string
    title: string
    url: string
    type: string
    course_id: string
    course_name?: string
  }
): Promise<FolderItem> => {
  const res = await apiClient.post<FolderItem>(`/api/folders/${folderId}/items`, item)
  return res.data
}

export const getItemsInFolder = async (folderId: string): Promise<FolderItem[]> => {
  const res = await apiClient.get<FolderItem[]>(`/api/folders/${folderId}/items`)
  return res.data
}

export const removeItemFromFolder = async (folderId: string, contentId: string): Promise<void> => {
  await apiClient.delete(`/api/folders/${folderId}/items/${contentId}`)
}

File: backend/app/routers/folders.py
from fastapi import APIRouter, HTTPException, Depends
from app.dependencies import get_current_user
from app.schemas.folder import FolderCreate, FolderUpdate, FolderResponse, FolderItemCreate, FolderItemResponse
from app.services import folder_service

router = APIRouter(prefix="/api/folders", tags=["Folders"])


@router.post("", response_model=FolderResponse, status_code=201)
def create_folder(body: FolderCreate, user=Depends(get_current_user)):
    return folder_service.create_folder(user["uid"], body)


@router.get("", response_model=list[FolderResponse])
def list_folders(user=Depends(get_current_user)):
    return folder_service.get_folders(user["uid"])


@router.patch("/{folder_id}", response_model=FolderResponse)
def rename_folder(folder_id: str, body: FolderUpdate, user=Depends(get_current_user)):
    try:
        return folder_service.rename_folder(user["uid"], folder_id, body)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.delete("/{folder_id}", status_code=204)
def delete_folder(folder_id: str, user=Depends(get_current_user)):
    try:
        folder_service.delete_folder(user["uid"], folder_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/{folder_id}/items", response_model=FolderItemResponse, status_code=201)
def add_item(folder_id: str, body: FolderItemCreate, user=Depends(get_current_user)):
    try:
        return folder_service.add_item(user["uid"], folder_id, body)
    except ValueError as e:
        detail = str(e)
        raise HTTPException(status_code=409 if "이미 저장" in detail else 404, detail=detail)


@router.get("/{folder_id}/items", response_model=list[FolderItemResponse])
def get_items(folder_id: str, user=Depends(get_current_user)):
    try:
        return folder_service.get_items(user["uid"], folder_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.delete("/{folder_id}/items/{content_id}", status_code=204)
def remove_item(folder_id: str, content_id: str, user=Depends(get_current_user)):
    try:
        folder_service.remove_item(user["uid"], folder_id, content_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
